pkg/transforms/mqtt.go

```diff
@@ -134,25 +136,36 @@ func (sender MQTTSender) MQTTSend(edgexcontext *appcontext.Context, params ...in
 		// We didn't receive a result
 		return false, errors.New("No Data Received")
 	}
+
+	exportData, err := util.CoerceType(params[0])
+	if err != nil {
+		return false, err
+	}
+
 	if !sender.client.IsConnected() {
 		edgexcontext.LoggingClient.Info("Connecting to mqtt server")
 		if token := sender.client.Connect(); token.Wait() && token.Error() != nil {
+			sender.setRetryData(edgexcontext, exportData)
 			return false, fmt.Errorf("Could not connect to mqtt server, drop event. Error: %s", token.Error().Error())
 		}
 		edgexcontext.LoggingClient.Info("Connected to mqtt server")
 	}
-	data, err := util.CoerceType(params[0])
-	if err != nil {
-		return false, err
-	}
-	token := sender.client.Publish(sender.topic, sender.opts.qos, sender.opts.retain, data)
-	// FIXME: could be removed? set of tokens?
+
+	token := sender.client.Publish(sender.topic, sender.opts.qos, sender.opts.retain, exportData)
 	token.Wait()
 	if token.Error() != nil {
+		sender.setRetryData(edgexcontext, exportData)
 		return false, token.Error()
 	}
-	edgexcontext.LoggingClient.Info("Sent data to MQTT Broker")
+
+	edgexcontext.LoggingClient.Debug("Sent data to MQTT Broker")
 	edgexcontext.LoggingClient.Trace("Data exported", "Transport", "MQTT", clients.CorrelationHeader, edgexcontext.CorrelationID)
 
 	return true, nil
 }
+
+func (sender MQTTSender) setRetryData(ctx *appcontext.Context, exportData []byte) {
+	if sender.persistOnError {
+		ctx.RetryData = exportData
```

Comment:
potential trace statement?

Reply:
I have a Trace in the runtime code when it is stored.